bluetooth/src/commonMain/kotlin/device/DeviceState.kt

```diff
@@ -0,0 +1,344 @@
+                        connect(deviceState)
+                    } else {
+                        deviceState.remain()
+                    }
+                }
+            }
+        }
+
+        fun connect(deviceState: DeviceState): suspend () -> DeviceState {
+            return if (deviceInfo.advertisementData.isConnectible) {
+                suspend {
+                    Connecting(deviceInfo, connectionManager)
+                }
+            } else {
+                deviceState.remain()
+            }
+        }
+    }
+
+    data class Disconnecting internal constructor(
+        override val deviceInfo: DeviceInfoImpl,
+        override val connectionManager: BaseDeviceConnectionManager
+    ) : DeviceState(deviceInfo, connectionManager), HandleAfterOldStateIsRemoved<DeviceState> {
+
+        override suspend fun afterOldStateIsRemoved(oldState: DeviceState) {
+            when (oldState) {
+                is Connecting, is Reconnecting, is Connected -> connectionManager.disconnect()
+                else -> {}
+            }
+        }
+    }
+
+    fun rssiDidUpdate(rssi: Int): suspend () -> DeviceState {
+        return updateDeviceInfo(deviceInfo.copy(rssi = rssi))
+    }
+
+    fun advertisementDataAndRssiDidUpdate(advertisementData: BaseAdvertisementData, rssi: Int): suspend () -> DeviceState {
+        return updateDeviceInfo(deviceInfo.copy(rssi = rssi, advertisementData = advertisementData))
+    }
+
+    private fun updateDeviceInfo(newDeviceInfo: DeviceInfoImpl): suspend () -> DeviceState {
+        return {
+            when (this) {
+                is Connected.NoServices -> copy(deviceInfo = newDeviceInfo)
+                is Connected.Idle -> copy(deviceInfo = newDeviceInfo)
+                is Connected.HandlingAction -> copy(deviceInfo = newDeviceInfo)
+                is Connected.Discovering -> copy(deviceInfo = newDeviceInfo)
+                is Connecting -> copy(deviceInfo = newDeviceInfo)
+                is Reconnecting -> copy(deviceInfo = newDeviceInfo)
+                is Disconnecting -> copy(deviceInfo = newDeviceInfo)
+                is Disconnected -> copy(deviceInfo = newDeviceInfo)
+            }
+        }
+    }
+
+    internal val didDisconnect = suspend {
+        Disconnected(deviceInfo, connectionManager)
+    }
+
+    internal val disconnecting = suspend {
+        Disconnecting(deviceInfo, connectionManager)
+    }
+}
+
+class Device internal constructor(
+    connectionSettings: ConnectionSettings,
+    private val initialDeviceInfo: DeviceInfoImpl,
+    connectionBuilder: BaseDeviceConnectionManager.Builder,
+    coroutineContext: CoroutineContext
+) : HotStateFlowRepo<DeviceState>(
+    coroutineContext = coroutineContext,
+    initialState = {
+        val deviceConnectionManager = connectionBuilder.create(connectionSettings, initialDeviceInfo.deviceHolder, it)
+        DeviceState.Disconnected(initialDeviceInfo, deviceConnectionManager)
+    }
+) {
+    val identifier: Identifier
```

Comment:
Can we expose `name` and `advertisementData` from info here as well?

Reply:
This fields are not dependent on state, but it's dynamics data, so I think it should be accessed from Device it self.